torch/_lazy/extract_compiled_graph.py

```diff
@@ -178,12 +179,21 @@ def extract_compiled_graph(model: fx.GraphModule, example_inputs) -> Callable:
     # sync the list of output tensors so the computation graph for these
     # tensors will be cached. Those computation graphs can be retrieved
     # by graph hash later.
-    lazy.sync_multi(lazy_out, [])
+    lazy.sync_multi(args_and_out, [])
 
     def optimized_mod(*args):
-        if len(lazy_out) == 0:
+        if len(args_and_out) == 0:
             return ()
         graph_input = graph_input_matcher(args)
-        return return_value_handler.duplicate_eager_tensors(computation.run_cached_graph(graph_hash, graph_input))
+        res = return_value_handler.duplicate_eager_tensors(computation.run_cached_graph(graph_hash, graph_input))
+
+        assert len(res) == len(args_and_out)
+        for i, arg in enumerate(args):
+            # only copy those tensors that get inplace updated
+            if arg is not res[i]:
```

Comment:
what does this 'is' check actually check?

Reply:
> what does this 'is' check actually check?

So we compiled a graph for `union(argument lazy tensors, returned lazy tensors)`. That graph will return (after duplication in case some tensors are duplicated) all the argument tensors at the beginning of the list. There are 2 cases
- some of those argument tensors may get inplace updated. In that case, the graph will return a new copy of the tensor. We need copy the content back to the argument
- other arguments may not get any inplace update. In that case, we can skip the copying.

Removing the 'is' check should also work but make things a bit slower.